ip6defrag/defrag.go

```diff
@@ -0,0 +1,227 @@
+			"(handcrafted? %d > %d)", fragOffset+uint16(len(in.Payload)), IPv6MaximumFragmentOffset)
+	}
+	return true, nil
+}
+
+func (f *fragmentList) insert(in gopacket.Packet) (gopacket.Packet, error) {
+	fragv6 := in.Layer(layers.LayerTypeIPv6Fragment).(*layers.IPv6Fragment)
+	fragOffset := fragv6.FragmentOffset * 8
+	if fragOffset >= f.Highest {
+		f.List.PushBack(in)
+	} else {
+		for e := f.List.Front(); e != nil; e = e.Next() {
+			packet, _ := e.Value.(gopacket.Packet)
+			frag := packet.Layer(layers.LayerTypeIPv6Fragment).(*layers.IPv6Fragment)
+			if fragv6.FragmentOffset <= frag.FragmentOffset {
+				debug.Printf("defrag: inserting frag %d before existing frag %d \n",
+					fragOffset, frag.FragmentOffset*8)
+				f.List.InsertBefore(in, e)
+				break
+			}
+		}
+	}
+	f.LastSeen = in.Metadata().Timestamp
+	fragLength := in.Layer(layers.LayerTypeIPv6).(*layers.IPv6).Length - 8 // need to conform
+	// After inserting the Fragment, we update the counters
+	if f.Highest < fragOffset+fragLength {
+		f.Highest = fragOffset + fragLength
+	}
+
+	f.Current = f.Current + fragLength
+
+	debug.Printf("defrag: insert ListLen: %d Highest:%d Current:%d\n",
+		f.List.Len(),
+		f.Highest, f.Current)
+
+	// Final Fragment ?
+	if fragv6.MoreFragments == false {
+		f.FinalReceived = true
+	}
+	// Ready to try defrag ?
+	if f.FinalReceived && f.Highest == f.Current {
+		return f.build(in)
+	}
+	return nil, nil
+}
+
+//need work
+func (f *fragmentList) build(in gopacket.Packet) (gopacket.Packet, error) {
+	var final []byte
+	var currentOffset uint16 = 0
+
+	debug.Printf("defrag: building the datagram \n")
+	for e := f.List.Front(); e != nil; e = e.Next() {
+		pack, _ := e.Value.(gopacket.Packet)
+		frag := pack.Layer(layers.LayerTypeIPv6Fragment).(*layers.IPv6Fragment)
+		ip := pack.Layer(layers.LayerTypeIPv6).(*layers.IPv6)
+		if frag.FragmentOffset*8 == currentOffset {
+			debug.Printf("defrag: building - adding %d\n", frag.FragmentOffset*8)
+			final = append(final, frag.Payload...)
+			currentOffset = currentOffset + ip.Length - 8
+		} else if frag.FragmentOffset*8 < currentOffset {
+			// overlapping fragment - let's take only what we need
+			startAt := currentOffset - frag.FragmentOffset*8
+			debug.Printf("defrag: building - overlapping, starting at %d\n",
+				startAt)
+			if startAt > ip.Length-8 {
+				return nil, fmt.Errorf("defrag: building - invalid fragment")
+			}
+			final = append(final, frag.Payload[startAt:]...)
+			currentOffset = currentOffset + frag.FragmentOffset*8
+		} else {
+			// Houston - we have an hole !
+			debug.Printf("defrag: hole found while building, " +
+				"stopping the defrag process\n")
+			return nil, fmt.Errorf("defrag: building - hole found")
+		}
+		debug.Printf("defrag: building - next is %d\n", currentOffset)
+	}
+	final_ipv6 := in.Layer(layers.LayerTypeIPv6).(*layers.IPv6)
+	final_frag := in.Layer(layers.LayerTypeIPv6Fragment).(*layers.IPv6Fragment)
+	// TODO recompute IP Checksum
```

Comment:
Hi, I am the code author. The reason I choose to build a packet and return it is that I use packet as input because I need get both IPv6 layer and IPv6 Fragmentation layer so I think it is more symmetrical to return the same data format as input. But if you think just return IPv6 layer is better, then I can change the code